xwp/pwa-wp
# Eliminate need for disabling navigation preload while adding navigation request caching strategy

Revisits #80/#67.

At the moment, when trying to enable a navigation request caching strategy (e.g. Network-First or Stale-While-Revalidate) you also have to disable navigation preload in order to successfully get the cached pages to be served instead of the offline page:

> <img width="800" alt="before-when-navigation-preload-enabled" src="https://user-images.githubusercontent.com/134745/60184587-99535580-9828-11e9-8051-41137dcaf354.png">

Navigation preload was developed to [improve performance of navigation requests](https://developers.google.com/web/updates/2017/02/navigation-preload) while the service worker is not booted.

The problem came down to incorrectly using `strategy.makeRequest()` when it should have been using `strategy.handle()`.

# Testing

Given a theme that has the following configuration for the service worker to enable a Network-First caching strategy for navigation requests:

```php
/*
 * Opt-in to Network-First caching strategy for navigation requests so that once you visit a page, you can go back to
 * that page even when you are offline, even when restarting the browser or reloading the page.
 */
add_filter(
	'wp_service_worker_navigation_caching_strategy',
	function() {
		return WP_Service_Worker_Caching_Routes::STRATEGY_NETWORK_FIRST;
	}
);

/*
 * Configure navigation request caching by limiting it to 50 entries and using the cache named 'pages'.
 * See https://github.com/xwp/pwa-wp/issues/176 for where something like this should be provided by default.
 */
add_filter(
	'wp_service_worker_navigation_caching_strategy_args',
	function( $args ) {
		$args['cacheName']                           = 'pages';
		$args['plugins']['expiration']['maxEntries'] = 50;
		return $args;
	}
);
```

## Before

Preload is not being used:

<img width="932" alt="after wp_service_worker_navigation_preload_false (disabled)" src="https://user-images.githubusercontent.com/134745/60184749-e505ff00-9828-11e9-8964-0540a5cc46be.png">

🚫 Offline page always displays even though it has been cached using the navigation caching strategy.

## After

Preload _is_ being used:

<img width="932" alt="after wp_service_worker_navigation_preload_true (enabled)" src="https://user-images.githubusercontent.com/134745/60185058-89884100-9829-11e9-81d5-de2acb7fc7df.png">

✅ Cached pages are displayed.

Additionally, an admin notice is added to help guide users to correct usage given that the `readme` was incorrect:

![image](https://user-images.githubusercontent.com/134745/60185727-b557f680-982a-11e9-9eaf-7b9fe71d6d03.png)

Even if the user does nothing, the navigation preload will be forced to be `true`, though a `_doing_it_wrong()` will be logged. Also, Site Health will inform of the navigation preload status. When it is being erroneously disabled:

<img width="877" alt="Screen Shot 2019-07-18 at 16 31 48" src="https://user-images.githubusercontent.com/134745/61493955-8260d880-a97a-11e9-9283-0090da1c39f6.png">

And when it is left enabled, with test passing:

<img width="826" alt="Screen Shot 2019-07-18 at 16 39 30" src="https://user-images.githubusercontent.com/134745/61493999-a3c1c480-a97a-11e9-9c38-d5e76a06dd23.png">

Lastly, when calling `\WP_Service_Workers::serve_request()` it will turn off `display_errors` to prevent debug messages from causing JS syntax errors in the service worker; this is what is done in the WP REST API as well.
